### ISSUE TITLE
Add JavaGenerator.printDaoConstructorAnnotations() to allow for custom annotations in generated DAO constructors

### ISSUE THREAD

**Issue by jklingsporn** [jOOQ/jOOQ#10801]
### Use case:

For my [side-project](https://github.com/jklingsporn/vertx-jooq/) I need to create custom constructor annotations for my generated DAOs. Unfortunately I have to overwrite the `protected void generateDao(TableDefinition table, JavaWriter out)` from the `JavaGenerator` entirely, as the generator does not allow customized DAO constructors:
https://github.com/jOOQ/jOOQ/blob/dddc1504b0095086d11d39b46cecca4d82a2f64d/jOOQ-codegen/src/main/java/org/jooq/codegen/JavaGenerator.java#L3902

### Possible solution you'd like to see:

Move the codeblock
```
if (generateSpringAnnotations())
            out.println("@%s", out.ref("org.springframework.stereotype.Repository"));
```
to a protected method like this:
```
protected void generateDAOConstructorAnnotations(TableDefinition table, JavaWriter out){
  if (generateSpringAnnotations())
            out.println("@%s", out.ref("org.springframework.stereotype.Repository"));
}
```
Edit: to be consistent with the generation of POJOs you could also put all constructor related code to it's own method like `protected void generateDaoDefaultConstructor(Definition tableOrUDT, JavaWriter out)`.
### Possible workarounds:

Copy the content of `protected void generateDao(TableDefinition table, JavaWriter out)` and customize. This method has the drawback of code duplication and does not benefit from new features or fixed bugs in that method in future jooq versions.

### Versions:

- jOOQ: 3.13/3.14

**Comment by lukaseder** [comment]
We have a `printClassAnnotations(JavaWriter, Definition, Mode)` method. In principle, I don't see why we couldn't have a `printConstructorAnnotations(JavaWriter, Definition, Mode)` method, too.

Of course, this whole concept of having overridable, protected `JavaGenerator` methods like these just allow for quick hacks, not well-designed custom logic. For example, generated tables or records have multiple constructors, and it doesn't always make sense to annotate them. Nor do we want to implement dozens of overridable methods, one for each constructor... In addition to this, we cannot reasonably evolve this API between minor releases, so we keep breaking it...

What do you need your annotations for, anyway?

**Comment by jklingsporn** [comment]
Guice or Dagger Inject-Annotations for example.

Lukas Eder <notifications@github.com> schrieb am Di., 27. Okt. 2020, 12:59:

**Comment by lukaseder** [comment]
> Guice or Dagger Inject-Annotations for example.

Can you give an example of what that would look like?

**Comment by jklingsporn** [comment]
https://github.com/jklingsporn/vertx-jooq/blob/adee3afb4f6820a4720a9327ff05bbc4e43fdf04/vertx-jooq-generate/src/test/java/generated/classic/jdbc/guice/vertx/tables/daos/SomethingDao.java#L29

Lukas Eder <notifications@github.com> schrieb am Di., 27. Okt. 2020, 13:23:

**Comment by lukaseder** [comment]
I see, thanks. So this is just a simple annotation without any arguments, that should be easy. I'll think of a viable solution to this, soon.

**Comment by lukaseder** [comment]
Well, I can't think of anything more sophisticated and reusable than your example. This won't be a documented feature - you know what you're doing... :)

**Comment by lukaseder** [comment]
Fixed for 3.15.0 and 3.14.4 (#11003).
